File: 5-longest-palindromic-substring/5-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        pair<string, int> longest;
        for (int i = 0; i < s.size(); i++) {
            string currOdd = getLongestPalindrome(s, i, i);
            int oddSize = currOdd.size();

            string currEven = i > 0 ? getLongestPalindrome(s, i-1, i) : "";
            int evenSize = currEven.size();
            
            int currLargest = max(oddSize, evenSize);
            if (currLargest > longest.second) {
                longest.first = currLargest == oddSize ? currOdd : currEven;
                longest.second = currLargest;
            }
        }
        return longest.first;
    }
    
    string getLongestPalindrome(string s, int left, int right) {
        // Compare left against right
        while (s[left] == s[right]) {
            if (left > 0 && right < s.size()) {
                left--;
                right++;
            } else {
                return s.substr(left, right-left+1);
            }
        }
        return s.substr(left+1, right-left-1);
    }
};

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) == 1:
            return s
        longest = ""
        for i in range(1, len(s)):
            even = self.isPalindrome(s, i-1,i)
            odd = self.isPalindrome(s, i, i)
            l = self.getLongest(even, odd)
            longest = self.getLongest(longest, l)
        return longest
        
    def getLongest(self, curr: str, toCompare: str) -> str:
        if len(curr) > len(toCompare):
            return curr
        return toCompare
        
    def isPalindrome(self, s: str, left: int, right: int) -> str:
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return s[left+1:right]
            